### PR TITLE
Bundle Iceberg reads into combined scan tasks

### PR DESCRIPTION
**Pull request by alexjo2144**
Reuse the bin packing done by the Iceberg table scan
planning by keeping small reads bundled together in
a single Split.

## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

> How would you describe this change to a non-technical end user or system administrator?

## Related issues, pull requests, and links

Fixes: https://github.com/trinodb/trino/issues/12162

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            this.delegatePageSources = ImmutableMap.copyOf(requireNonNull(delegatePageSourcesByPath, "delegatePageSources is null"));
+            this.pageSourceIterator = delegatePageSourcesByPath.values().iterator();
+            this.currentPageSource = pageSourceIterator.next();
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegatePageSources.values().stream()
```
Comment (electrum) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java:68]: This recomputes the values for every call. Instead, we can compute as we go, since the stats for completed page sources will not change. See `io.trino.plugin.raptor.legacy.util.ConcatPageSource`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                .mapToLong(ConnectorPageSource::getMemoryUsage)
+                .sum();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        IOException closeException = null;
```
Comment (electrum) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java:122]: Simply add them all to a `com.google.common.io.Closer` which handles this automatically.

**Review by electrum — CHANGES_REQUESTED** [review]
Doing this for UPDATE and DELETE makes the code complicated and hard to follow, and the amount of test coverage is not clear. After we have `MERGE`, we'll eventually replace the existing UPDATE and DELETE implementations to use it, so we won't need `UpdatablePageSource` anymore.

We can support this only for reads like we do in Raptor by only packing for reads. The table handle can have a flag to know if it's a write (it already does for UPDATE, just needs one for DELETE). Then the split source can skip packing for writes. On the page source provider side, skip creating the combined page source if there is only one split.

I suspect almost all of this complexity goes away if we only support reads.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplit.java [hunk 3]
```diff
@@ -72,89 +53,175 @@ public boolean isRemotelyAccessible()
...
-                .put("path", path)
-                .put("start", start)
-                .put("length", length)
-                .buildOrThrow();
+        return new Builder();
     }
 
-    @Override
-    public long getRetainedSizeInBytes()
+    public static class Builder
```
Comment (electrum) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplit.java:79]: No need for a builder class. This is basically just an `ImmutableList.Builder`.

**Review by electrum — COMMENTED** [review]
With the recently added support for multiple handles, no need to make changes to `IcebergSplit`. You can add a new class `CombinedIcebergSplit` that contains `List<IcebergSplit>`. Then just do a simple `instanceof` check in `IcebergPageSourceProvider` and turn the existing `createPageSource()` method into a private helper method that you call in a loop.

**Comment by alexjo2144** [comment]
> Doing this for UPDATE and DELETE makes the code complicated

Yeah, pretty much all of it is just there to get the rows back to the PageSource that generated them. I can try going with the original Split handling for update and delete.

Alternatively we could write one new data file and one delete file per batch, and not worry about delegating at all. The complexity there is just that the delete file needs to be sorted by the data file path.

**Comment by findepi** [comment]
> we could write one new data file and one delete file per batch

That sounds undesirable (more data to read at read-time, than if one deletion file per one input data file)

I like the idea to keep this simple, and apply it to read queries only. We may always revisit this later if needed.

**Comment by sopel39** [comment]
So with packing splits there is less FileScanTasks? Can single FileScanTask read multiple files? Why FileScanTask is a bottleneck? I suppose the same amount of IO needs to happen.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import io.trino.spi.connector.ConnectorSplit;
+
+import java.util.Collection;
+import java.util.List;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class CombinedIcebergSplit
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:30]: Did you try assigning split weight (as in hive) instead?

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:30]: How do weighted split work?
does it allow splits to be combined to avoid scheduling overhead?

**Comment by findepi** [comment]
cc @losipiuk

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java (hunk 4 above)
Comment (sopel39) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:30]: > How do weighted split work?

Check https://github.com/trinodb/trino/pull/9059. It's pretty well described there.

> does it allow splits to be combined to avoid scheduling overhead?

Scheduling overhead itself is minimal. Split queues might be too short if there is a lot of splits, but that's what weighted split scheduling solves

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    private final Iterator<ConnectorPageSource> pageSourceIterator;
+    private final Closer closer = Closer.create();
+
+    private ConnectorPageSource currentPageSource;
+    private long completedBytes;
+    private long readTimeNanos;
+
+    public IcebergCombinedPageSource(List<ConnectorPageSource> delegatePageSources)
+    {
+        if (delegatePageSources.isEmpty()) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java:48]: for simplicity we may  `checkArgument(!delegatePageSources.isEmpty()` to avoid dead code

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            this.pageSourceIterator = delegatePageSources.iterator();
+            this.currentPageSource = pageSourceIterator.next();
+            this.delegatePageSources.forEach(closer::register);
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return completedBytes + currentPageSource.getCompletedBytes();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java:64]: completedBytes -> previouslyCompletedBytes 
(but, this pattern won't sound well for readTimeNanos. thoughts?)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @Override
+    public OptionalLong getCompletedPositions()
+    {
+        return delegatePageSources.stream()
+                .map(ConnectorPageSource::getCompletedPositions)
+                .reduce(OptionalLong.of(0), (accumulatedPositions, newPositions) -> {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java:72]: Why is it so different than getCompletedBytes ?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    @Override
+    public Page getNextPage()
+    {
+        if (currentPageSource.isFinished()) {
+            completedBytes += currentPageSource.getReadTimeNanos();
+            readTimeNanos += currentPageSource.getReadTimeNanos();
+            try {
+                currentPageSource = pageSourceIterator.next();
+            }
+            catch (NoSuchElementException e) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java:101]: use `hasNext`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 9]
```diff
@@ -336,7 +363,7 @@ public ConnectorPageSource createPageSource(
 
         return new IcebergPageSource(
                 tableSchema,
-                icebergColumns,
+                columns,
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:366]: possible to retract some renames?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java [hunk 10]
```diff
@@ -178,74 +173,177 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHan
...
-                        scanTask.file().upperBounds(),
-                        scanTask.file().nullValueCounts())) {
-                    continue;
-                }
+        @Override
+        public void close()
+                throws IOException
+        {
+            this.fileScanTaskIterable.close();
+            this.fileScanTaskIterator.close();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:271]: curious, Why you need to close both?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java [hunk 11]
```diff
@@ -178,74 +173,177 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHan
...
+            this.fileScanTaskIterator = fileScanTaskIterable.iterator();
+            this.next = getNext();
+        }
 
-            Schema fileSchema = scanTask.spec().schema();
-            Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(scanTask);
+        @Override
+        public boolean hasNext()
+        {
+            return next != null;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:255]: `com.google.common.base.AbstractIterator` is your friend

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java [hunk 12]
```diff
@@ -178,74 +173,177 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHan
...
+        {
+            if (!fileScanTaskIterator.hasNext()) {
+                return null;
+            }
+
+            return splitForFileScanTask(fileScanTaskIterator.next()).orElse(null);
+        }
+    }
+
+    private Optional<IcebergSplit> splitForFileScanTask(FileScanTask scanTask)
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:284]: Can you please extract some refactorings?

This shows up as a single huge change, but i guess most code is just being moved around.
Please make it apparent.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java [hunk 13]
```diff
@@ -178,74 +173,177 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHan
...
-        this.combinedScanIterable = CloseableIterable.empty();
-        this.fileScanIterator = Collections.emptyIterator();
+        this.splitIterator = CloseableIterator.empty();
     }
 
     @Override
     public boolean isFinished()
     {
-        return fileScanIterator != null && !fileScanIterator.hasNext();
+        return splitIterator != null && !splitIterator.hasNext();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:346]: finish has `this.splitIterator = CloseableIterator.empty();`

let's decide whether `splitIterator` is nullable and simplify finish or isfinished

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java [hunk 14]
```diff
@@ -80,6 +81,7 @@ public IcebergTableHandle(
             @JsonProperty("tableLocation") String tableLocation,
             @JsonProperty("storageProperties") Map<String, String> storageProperties,
             @JsonProperty("retryMode") RetryMode retryMode,
+            @JsonProperty("isUpdateOrDelete") boolean isUpdateOrDelete,
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java:84]: Sounds like something engine could (should?) provide directly to `ConnectorSplitManager`.

cc @losipiuk

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergNodeLocalDynamicSplitPruning.java [hunk 15]
```diff
@@ -153,7 +153,7 @@ private static void writeOrcContent(File file)
 
     private static ConnectorPageSource createTestingPageSource(HiveTransactionHandle transaction, IcebergConfig icebergConfig, HiveConfig hiveConfig, File outputFile, DynamicFilter dynamicFilter)
     {
-        IcebergSplit split = new IcebergSplit(
+        CombinedIcebergSplit split = new CombinedIcebergSplit(ImmutableList.of(new IcebergSplit(
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergNodeLocalDynamicSplitPruning.java:156]: redundant change?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 16]
```diff
@@ -229,8 +228,43 @@ public ConnectorPageSource createPageSource(
...
+        }
+
+        if (connectorSplit instanceof CombinedIcebergSplit) {
+            CombinedIcebergSplit icebergSplit = (CombinedIcebergSplit) connectorSplit;
+            return new IcebergCombinedPageSource(
+                    icebergSplit.getEntries().stream()
+                            .map(entry -> createIcebergPageSource(session, table, entry, tableSchema, hdfsContext, fileIO, icebergColumns, nameMapping, effectivePredicate))
+                            .collect(toImmutableList()));
+        }
+        else if (connectorSplit instanceof IcebergSplit) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:248]: redundant `else`

**Review by findepi — COMMENTED** [review]
skimming

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java (hunk 4 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:30]: I like that provided this solves the problem.

@alexjo2144 can you please make another version of the improvement, based on weighted splits?

**Review by findepi — CHANGES_REQUESTED** [review]
Pending discussion https://github.com/trinodb/trino/pull/12359#discussion_r876656641

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java (hunk 4 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:31]: Here's combined splits + split weights:
```
trino> select count(*) from iceberg.iceberg_tpch_sf1000_orc.lineitem_small_files;
   _col0    
------------
 5999989709 
(1 row)

Query 20220519_200949_00011_empm9, FINISHED, 6 nodes
Splits: 7,469 total, 7,469 done (100.00%)
24.89 [6B rows, 7.25TB] [241M rows/s, 298GB/s]
```

Comment (sopel39) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:31]: > So still faster than we have on master, but there seems to still be some extra overhead from the number of splits being so high.

How do you assign split weight? Is it similar to Hive connector approach?

> but there seems to still be some extra overhead from the number of splits being so high.

That is an assumption. It can either be:
* some kind of overhead
* workers still get starved as queue size is not sufficient even with weighted splits. You can actually try increasing split queue size (`node-scheduler.max-splits-per-node`, `node-scheduler.max-pending-splits-per-task`) to see if it helps. Is Iceberg generating splits fast enough? Check `query.schedule-split-batch-size`.

I would like to learn where the overhead comes from since we already have a mechanism (weighted splits) that should address the issue of multiple small splits. It would be better to fix/tune it rather than add yet another mechanism.

If the conclusion is that the overhead is X and it can be only addressed in Iceberg with merging splits, that's fine. However, if we start packing splits together for other connectors we should know where is the overhead.

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:31]: > I would like to learn where the overhead comes from since we already have a mechanism (weighted splits) that should address the issue of multiple small splits. It would be better to fix/tune it rather than add yet another mechanism.

That's a good point

> If the conclusion is that the overhead is X and it can be only addressed in Iceberg with merging splits, that's fine.

or we can even think of making engine-side equivalent of these changes, so connectors don't have to do this

Comment (sopel39) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CombinedIcebergSplit.java:31]: > or we can even think of making engine-side equivalent of these changes, so connectors don't have to do this

Sure, but then you probably don't need weighted split scheduling. Also, large splits are not neccecerly good because they provide less granularity of work which leads to work skewness (especially when there are more nodes)

**Comment by alexjo2144** [comment]
Bundling did not end up having a significant improvement over using SplitWeights. Replacing this PR with https://github.com/trinodb/trino/pull/12579
